### ISSUE TITLE
proxyForInterface is confused about the difference between 'implements' and 'provides'

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3646]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)| @glyph reported|
|-|-|
|Trac ID|trac#3646|
|Type|enhancement|
|Created|2009-02-06 11:29:05Z|
|Branch|https://github.com/twisted/twisted/tree/proxyforinterface-implements-3646|

`proxyForInterface` declares that the type it creates `directlyProvides` the interface that is passed.  This is not true: the type _implements_ the given interface.

This example demonstrates the problem:

```
#!py
from zope.interface import Interface, implements
from zope.interface.verify import verifyObject
from twisted.python.components import proxyForInterface

class IFoo(Interface):
    def bar():
        ""

    def baz():
        ""


class Implementation(object):
    implements(IFoo)
    def bar(self):
        print 'bar'
    def baz(self):
        print 'baz'

class Wrapper(proxyForInterface(IFoo)):
    def bar(self):
        print 'nope, not bar.'

candidate = Wrapper(Implementation())
print verifyObject(IFoo, candidate)
```

Users can work around the more practically annoying half of this problem by declaring `implements(IFoo)` in `Wrapper`.  However, there's another problem.  In the above example, this should raise an exception, but instead prints 'True':
```
#!py
print verifyObject(IFoo, proxyForInterface(IFoo))
```

The following patch should fix it:
```
#!diff
Index: twisted/python/components.py
===================================================================
--- twisted/python/components.py	(revision 26150)
+++ twisted/python/components.py	(working copy)
@@ -34,7 +34,7 @@
 import warnings
 
 # zope3 imports
-from zope.interface import directlyProvides, interface, declarations
+from zope.interface import interface, declarations
 from zope.interface.adapter import AdapterRegistry
 
 
@@ -348,7 +348,7 @@
         contents[name] = _ProxyDescriptor(name, originalAttribute)
     proxy = type("(Proxy for %s)"
                  % (reflect.qual(iface),), (object,), contents)
-    directlyProvides(proxy, iface)
+    declarations.classImplements(proxy, iface)
     return proxy
 
 
```

<details><summary>Searchable metadata</summary>

```
trac-id__3646 3646
type__enhancement enhancement
reporter__glyph glyph
priority__normal normal
milestone__ 
branch__branches_proxyforinterface_implements_3646 branches/proxyforinterface-implements-3646
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1233919745000000 1233919745000000
changetime__1234662335000000 1234662335000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

(In [26198]) Branching to 'proxyforinterface-implements-3646'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

(In [26199]) Fix proxyForInterface to make classes implement, not provide, the proxied interface

refs [#3646](https://github.com/twisted/twisted/issues/3646)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun removed owner|
|-|-|

[Build Results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/proxyforinterface-implements-3646)

Please review.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve set owner to @exarkun|
|-|-|

Great, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set status to `closed`|
|-|-|

(In [26222]) Merge proxyforinterface-implements-3646 

Author: exarkun
Reviewer: therve
Fixes: [#3646](https://github.com/twisted/twisted/issues/3646)

Change `proxyForInterface` so that the class objects returned by it implement the interface
being proxied, rather than providing it.  This is what one would expect from a class claiming
to act as a proxy.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
